File: src/2594.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    long long repairCars(vector<int>& ranks, int cars) {
        long long retVal = 0;

        // The minimum possible time is 1
        // The maximum possible time is when the slowest mechanic (highest rank) repairs all cars
        long long left = 1;
        long long right = 1LL * cars * cars * ranks[0];
        while (left < right) {
            long long middle = left + (right - left) / 2;

            // Calculate the number of cars that can be repaired in 'middle' time by all mechanics.
            long long carsRepaired = 0;
            for (auto rank : ranks) {
                carsRepaired += sqrt(1.0 * middle / rank);
            }

            // If the total cars repaired is less than required, increase the time. Otherwise, try a smaller time.
            if (carsRepaired < cars) {
                left = middle + 1;

            } else {
                right = middle;
            }
        }
        retVal = left;

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<int> ranks;
        int cars;
    };
    vector<subject> testData{{{4, 2, 3, 1}, 10}, {{5, 1, 8}, 6}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: ranks = [4,2,3,1], cars = 10
     *  Output: 16
     *
     *  Input: ranks = [5,1,8], cars = 6
     *  Output: 16
     */

    Solution cSolution;
    long long answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: ranks = [";
        for (long unsigned int j = 0; j < testData[i].ranks.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << testData[i].ranks[j];
        }
        cout << "], cars = " << testData[i].cars << "\n";

        answer = cSolution.repairCars(testData[i].ranks, testData[i].cars);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/2594.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def repairCars(self, ranks: List[int], cars: int) -> int:
        retVal = 0

        # The minimum possible time is 1
        # The maximum possible time is when the slowest mechanic (highest rank) repairs all cars
        left = 1
        right = cars * cars * ranks[0]
        while left < right:
            middle = (left + right) // 2

            carsRepaired = sum(int((middle / rank) ** 0.5) for rank in ranks)

            # If the total cars repaired is less than required, increase the time.  Otherwise, try a smaller time.
            if carsRepaired < cars:
                left = middle + 1
            else:
                right = middle
        retVal = left

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for ranks, cars in zip([[4, 2, 3, 1], [5, 1, 8]], [10, 6]):
            # /* Example
            #  *  Input: ranks = [4,2,3,1], cars = 10
            #  *  Output: 16
            #  *
            #  *  Input: ranks = [5,1,8], cars = 6
            #  *  Output: 16
            #  */
            logging.info("Input: ranks = %s, cars = %d", ranks, cars)

            retVal = pSolution.repairCars(ranks, cars)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
